### PR TITLE
Improved Explodes trait and how artillery explodes

### PR DESCRIPTION
**Pull request by reaperrr**
Added `LoadedChance` to `Explodes`, changed how RA/TD artillery explodes.

See individual commits for more detailed descriptions.

### CODE REVIEW

**Comment by reaperrr** [comment]
I don't want to make that call on my own, but I think this should go into next playtest/release.

**Comment by pchote** [comment]
> Reduced TD UnitExplode damage to 1/10 of old value.

This seems a bit drastic.  What's the balance motivation for such a large change?

**Comment by reaperrr** [comment]
> This seems a bit drastic. What's the balance motivation for such a large change?

That this explosion currently isn't used at all in TD since #8303 removed it from artillery precisely because it dealt way too much (splash) damage.

File: OpenRA.Mods.Common/Traits/Explodes.cs [hunk 1]
```diff
@@ -14,14 +14,22 @@
...
 {
+	[Desc("This actor explodes when killed.")]
 	public class ExplodesInfo : ITraitInfo
 	{
-		[WeaponReference]
+		[WeaponReference, Desc("Weapon to use for explosion if ammo/payload is loaded.")]
 		public readonly string Weapon = "UnitExplode";
-		[WeaponReference]
+
+		[WeaponReference, Desc("Weapon to use for explosion if no ammo/payload.")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Explodes.cs]: "`if no ammo/payload` ... is loaded."?

**Comment by abcdefg30** [comment]
Some thing about `LoadedChance` seems wrong to me...
If you set it to 0, the Weapon explosion will always occur.

**Comment by Unit158** [comment]
As it should. Loaded chance only changes the likeliness of a different weapon being used.

**Comment by abcdefg30** [comment]
`[Desc("Chance that the explosion will use Weapon if the actor has ammo/payload.")]`
Sounds to me that `Weapon` shouldn't be used if the chance is 0.

File: OpenRA.Mods.Common/Traits/Explodes.cs [hunk 2]
```diff
@@ -60,7 +68,8 @@ public void Killed(Actor self, AttackInfo e)
 		string ChooseWeaponForExplosion(Actor self)
 		{
 			var shouldExplode = self.TraitsImplementing<IExplodeModifier>().All(a => a.ShouldExplode(self));
-			return shouldExplode ? info.Weapon : info.EmptyWeapon;
+			var useFullExplosion = self.World.SharedRandom.Next(100) > info.LoadedChance;
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/Explodes.cs]: > should be <= :)

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/Explodes.cs [hunk 3]
```diff
@@ -14,14 +14,22 @@
...
-		[WeaponReference]
+		[WeaponReference, Desc("Weapon to use for explosion if ammo/payload is loaded.")]
 		public readonly string Weapon = "UnitExplode";
-		[WeaponReference]
+
+		[WeaponReference, Desc("Weapon to use for explosion if no ammo/payload is loaded.")]
 		public readonly string EmptyWeapon = "UnitExplode";
 
+		[Desc("Chance that the explosion will use Weapon if the actor has ammo/payload.")]
+		public readonly int LoadedChance = 100;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Explodes.cs]: This changes the previous default.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Explodes.cs]: Or wait probably not. This is just confusingly complicated.

**Comment by Mailaender** [comment]
Stuff still explodes properly. :white_check_mark:

**Comment by obrakmann** [comment]
The artillery in RA now makes the 'artillery firing sound' when it explodes. TD is fine.

**Comment by reaperrr** [comment]
Updated.

- Reverted RA changes to `UnitExplode`, since it's the default explosion used by some traits (crashing aircraft, for example). This was not necessary for TD since it uses custom `HeliCrash` and `HeliExplode` for helicopters.
- Introduced an `ArtilleryExplosion` to RA instead, which is now used as `Weapon` when artillery explodes. 
- Artillery `EmptyWeapon` is now `UnitExplodeSmall` like for all other vehicles.

**Comment by obrakmann** [comment]
Looks ok now. :+1:

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/82f45c2533da91fd4efb8e4f3de023331c2ec8c4%5E...82f45c2533da91fd4efb8e4f3de023331c2ec8c4)

**Pull request merged**